File: src/main/java/sujung/linkedlist/RemoveDuplicatesFromSortedList2.java
package sujung.linkedlist;

public class RemoveDuplicatesFromSortedList2 {

    private static class ListNode {
        int val;
       ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode deleteDuplicates(ListNode head) {
        ListNode answer = new ListNode();
        ListNode pre = head;
        ListNode post = answer;
        post.next = pre;

        while(pre != null) {
            while (pre.next != null && pre.val == pre.next.val) {
                pre = pre.next;
            }
            if(post.next != pre) {
                post.next = pre.next;
                pre = pre.next;
            } else {
                pre = pre.next;
                post = post.next;
            }
        }
        return answer.next;
    }
}

File: src/main/java/thkim/linkedlist/RemoveDuplicatesFromSortedList2.kt
package thkim.linkedlist

fun deleteDuplicates2(head: ListNode?): ListNode? {
    var head = head

    // head 가 마지막 값일 때.
    if (head?.next == null) return head

    return if (head.`val` != head.next!!.`val`) {
        head.next = deleteDuplicates2(head.next) // 다음 노드 결정.
        head
    } else {
        // 중복 값일 때 현재 값에 다음 값을 지정
        while (head!!.next != null && head.`val` == head.next!!.`val`) head = head.next
        deleteDuplicates2(head.next)
    }
}